Mathlib/Tactic/Ring/RingNF.lean

```diff
@@ -119,12 +119,10 @@ theorem add_neg {R} [Ring R] (a b : R) : a + -b = a - b := (sub_eq_add_neg ..).s
 theorem nat_rawCast_0 : (Nat.rawCast 0 : R) = 0 := by simp
 theorem nat_rawCast_1 : (Nat.rawCast 1 : R) = 1 := by simp
 theorem nat_rawCast_2 [Nat.AtLeastTwo n] : (Nat.rawCast n : R) = OfNat.ofNat n := rfl
-theorem int_rawCast_1 {R} [Ring R] : (Int.rawCast (.negOfNat 1) : R) = -1 := by
-  simp [Int.negOfNat_eq]
-theorem int_rawCast_2 {R} [Ring R] [Nat.AtLeastTwo n] :
-    (Int.rawCast (.negOfNat n) : R) = -OfNat.ofNat n := by
-  simp [Int.negOfNat_eq, OfNat.ofNat]
-theorem rat_rawCast_2 {R} [DivisionRing R] : (Rat.rawCast n d : R) = n / d := by simp
+theorem int_rawCast_pos {R} [Ring R] : (Int.rawCast (.ofNat n) : R) = Nat.rawCast n := by simp
+theorem int_rawCast_neg {R} [Ring R] : (Int.rawCast (.negOfNat n) : R) = -Nat.rawCast n := by simp
+theorem rat_rawCast {R} [DivisionRing R] :
+    (Rat.rawCast n d : R) = Int.rawCast n / Nat.rawCast d := by simp
```

Comment:
The docstring in the other file says that `Int.rawCast (.ofNat n)` is not allowed to ever appear; so arguably `rat_rawCast` is a bad lemma because it introduces this illegal form, even if only temporarily.

I think the version I had that preserved the invariant was a little clearer.

Reply:
We're taking things out of raw cast form here, so it's fine to break those invariants. (More precisely we are temporarily weakening one of the invariants to allow the `Int.rawCast (.ofNat n)` form.) The main reason I like this version better is that it's more orthogonal, we only need one case each for negation and division and the int cast. Otherwise we need a cartesian product of cases for all the ways that you can build literals.

Reply:
Just for comparison, my proposal is
```suggestion
theorem int_rawCast_neg {R} [Ring R] : (Int.rawCast (.negOfNat n) : R) = -Nat.rawCast n := by simp
theorem rat_rawCast_pos {R} [DivisionRing R] :
    (Rat.rawCast (.ofNat n) d : R) = Nat.rawCast n / Nat.rawCast d := by simp
theorem rat_rawCast_neg {R} [DivisionRing R] :
    (Rat.rawCast (.negOfNat n) d : R) = Int.rawCast (.negOfNat n) / Nat.rawCast d := by simp
```
which has two benefits:
* It never transitions through an invalid use of `Int.rawCast`
* It doesn't accept invalid uses of `Rat.rawCast`

This is hardly a cartesian explosion, since we only have three types here anyway.

Reply:
Bhavik posted the following example on Zulip:
```
import Mathlib.Data.Real.Basic
example {n : ℝ} : (n + 1 / 2) ^ 2 * (n + 1 + 1 / 3) ≤ (n + 1 / 3) * (n + 1) ^ 2 := by ring_nf; sorry
```
On 86251e6, the remaining goal displays as `⊢ 1 / 3 + n * (OfNat.ofNat 19 / 12) + n ^ 2 * (OfNat.ofNat 7 / 3) + n ^ 3 ≤ 1 / 3 + n * (OfNat.ofNat 5 / 3) + n ^ 2 * (OfNat.ofNat 7 / 3) + n ^ 3` in the infoview. Applying Eric's suggestion, it is `⊢ 1 / 3 + n * (19 / 12) + n ^ 2 * (7 / 3) + n ^ 3 ≤ 1 / 3 + n * (5 / 3) + n ^ 2 * (7 / 3) + n ^ 3`

Reply:
(Note, regarding the two simp lemma strategies, I don't really care that much about it. Use whichever one you prefer.)

Reply:
@collares, do you have any ideas for how to write a test for that? It seems the one I added doesn't spot the issue.

Reply:
I ended up using `guard_mgs` with `trace_state`